crowbar_framework/app/models/service_object.rb

```diff
@@ -1205,31 +1306,40 @@ def apply_role(role, inst, in_queue)
           end
 
           old_nodes.each do |node_name|
-            # Don't add deleted nodes to the run order
+            # Don't add deleted nodes to the run order, they clearly won't have the old role
             if NodeObject.find_node_by_name(node_name).nil?
               @logger.debug "skipping deleted node #{node_name}"
               next
             end
 
+            # An old node that is not in the new deployment, drop it
             if new_nodes.nil? or !new_nodes.include?(node_name)
               @logger.debug "remove node #{node_name}"
               pending_node_actions[node_name] = { :remove => [], :add => [] } if pending_node_actions[node_name].nil?
+
               pending_node_actions[node_name][:remove] << role_name
+
+              # FIXME: are remove_roles a way how to tell us to do some post-cleanup?
```

Comment:
It's a way to "de-configure" things on the node when a role is not used anymore for that node. For instance, stopping a service, or removing packages.